### PR TITLE
Tidy-up of datatypes checking

### PR DESCRIPTION
**Pull request by cgsandford**
Follow-up from #978 

Added bounds checking to `check_datatypes` and made functional unit tests for `check_cube_not_float64`.  Last commit is code rearrangements with no functional changes - suggest reviewers look at all but the last commit for a clear picture of what has changed.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/metadata/check_datatypes.py [hunk 1]
```diff
@@ -217,3 +187,43 @@ def _check_units_and_dtype(obj, units, dtype):
         return False
 
     return True
+
+
+def check_time_coordinate_metadata(cube):
```
Comment (cgsandford) [inline lib/improver/metadata/check_datatypes.py]: Function moved without changes.

File: lib/improver/tests/metadata/test_check_datatypes.py [hunk 2]
```diff
@@ -259,26 +253,53 @@ def setUp(self):
...
         self.assertTrue(result)
 
     def test_fail_coord(self):
         """Test return value for non-compliant coordinate"""
-        result = enforce._check_units_and_dtype(self.coord, 'm', np.int32)
+        result = _check_units_and_dtype(self.coord, 'm', np.int32)
         self.assertFalse(result)
 
 
+class Test_check_time_coordinate_metadata(IrisTest):
```
Comment (cgsandford) [inline lib/improver/tests/metadata/test_check_datatypes.py]: Test class moved without changes (to remain consistent with order of functions in module).

File: lib/improver/tests/metadata/test_check_datatypes.py [hunk 3]
```diff
@@ -46,28 +48,44 @@ class Test_check_cube_not_float64(IrisTest):
...
+        +-------------------------+-------------+
+        | projection_y_coordinate | np.float32  |
+        +-------------------------+-------------+
+        | time                    | np.int64    |
+        +-------------------------+-------------+
+        | forecast_reference_time | np.int64    |
+        +-------------------------+-------------+
+        | forecast_period         | np.int32    |
+        +-------------------------+-------------+
+        """
```
[the remark this answers is not in the document]
Comment (cgsandford) [reply lib/improver/tests/metadata/test_check_datatypes.py]: @benfitzpatrick What do you think about adding a section on datatypes to the code style guide?  Can do so if you approve.

[the remark this answers is not in the document]
Comment (benfitzpatrick) [reply lib/improver/tests/metadata/test_check_datatypes.py]: I think that's a great idea

[the remark this answers is not in the document]
Comment (cgsandford) [reply lib/improver/tests/metadata/test_check_datatypes.py]: OK, I've added a section on "datatypes" under "General", and moved up the "avoiding float64" to join it.

**Pull request merged**